src/NHibernate/Linq/Visitors/TransparentIdentifierRemovingExpressionVisitor.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using System.Reflection;
+using Remotion.Linq;
+using Remotion.Linq.Clauses.Expressions;
+using Remotion.Linq.Parsing;
+using MemberBinding = Remotion.Linq.Parsing.ExpressionVisitors.MemberBindings.MemberBinding;
+
+namespace NHibernate.Linq.Visitors
+{
+	// Copied from Relinq and added a fallback for comparing two member info by DeclaringType and Name
+	/// <summary>
+	/// Replaces expression patterns of the form <c>new T { x = 1, y = 2 }.x</c> (<see cref="MemberInitExpression"/>) or 
+	/// <c>new T ( x = 1, y = 2 ).x</c> (<see cref="NewExpression"/>) to <c>1</c> (or <c>2</c> if <c>y</c> is accessed instead of <c>x</c>).
+	/// Expressions are also replaced within subqueries; the <see cref="QueryModel"/> is changed by the replacement operations, it is not copied. 
+	/// </summary>
+	internal sealed class TransparentIdentifierRemovingExpressionVisitor : RelinqExpressionVisitor
+	{
+		public static Expression ReplaceTransparentIdentifiers(Expression expression)
+		{
+			Expression expressionBeforeRemove;
+			Expression expressionAfterRemove = expression;
+
+			// Run again and again until no replacements have been made.
+			do
+			{
+				expressionBeforeRemove = expressionAfterRemove;
+				expressionAfterRemove = new TransparentIdentifierRemovingExpressionVisitor().Visit(expressionAfterRemove);
+			} while (expressionAfterRemove != expressionBeforeRemove);
+
+			return expressionAfterRemove;
+		}
+
+		private TransparentIdentifierRemovingExpressionVisitor()
+		{
+		}
+
+		protected override Expression VisitMember(MemberExpression memberExpression)
+		{
+			var memberBindings = GetMemberBindingsCreatedByExpression(memberExpression.Expression);
+			if (memberBindings == null)
+				return base.VisitMember(memberExpression);
+
+			var matchingAssignment = memberBindings
+			                         .Where(binding => binding.MatchesReadAccess(memberExpression.Member))
+			                         .LastOrDefault();
+
+			// Added logic: In some cases (e.g OData), the member can be from a different derived class, in such case
+			// we need to check the member DeclaringType instead of ReflectedType
+			if (matchingAssignment == null && memberExpression.Expression.NodeType == ExpressionType.MemberInit)
```

Comment:
The only way to test this was to include and use `Microsoft.AspNetCore.OData` package as this cannot be reproduced by creating classes in code.